File: 2020/Robot Testing/2019RobotFieldcentric/src/main/java/frc/robot/Robot.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package frc.robot;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

import edu.wpi.cscore.CvSink;
import edu.wpi.cscore.CvSource;
import edu.wpi.cscore.UsbCamera;

import edu.wpi.first.cameraserver.*;
import org.opencv.core.Mat;
import org.opencv.imgproc.Imgproc;

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to each mode, as described in the TimedRobot
 * documentation. If you change the name of this class or the package after
 * creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot
{
private final XboxController DriverController = new XboxController(0);
  private final XboxController ElevatorController = new XboxController(1);

  private SwerveDriveTrain driveTrain;
  private Elevator elevator;

  private boolean isElevGoing = false;
  private static final String kDefaultAuto = "Default";
  private static final String kCustomAuto = "My Auto";
  private String m_autoSelected;
  private final SendableChooser<String> m_chooser = new SendableChooser<>();

  /**
   * This function is run when the robot is first started up and should be
   * used for any initialization code.
   */
  @Override
  public void robotInit() {

    new Thread(() -> {
      UsbCamera camera = CameraServer.getInstance().startAutomaticCapture();
      camera.setResolution(320, 240);
      
      CvSink cvSink = CameraServer.getInstance().getVideo();
      CvSource outputStream = CameraServer.getInstance().putVideo("Video_1", 160, 120);
      
      Mat source = new Mat();
      Mat output = new Mat();
      
      while(!Thread.interrupted()) {
          cvSink.grabFrame(source);
          Imgproc.cvtColor(source, output, Imgproc.COLOR_BGR2GRAY);
          outputStream.putFrame(output);
      }
    }).start();

    m_chooser.setDefaultOption("Default Auto", kDefaultAuto);
    m_chooser.addOption("My Auto", kCustomAuto);
    SmartDashboard.putData("Auto choices", m_chooser);
    driveTrain = SwerveDriveTrain.getInstance();
    elevator = Elevator.getInstance();
  }

  /**
   * This function is called every robot packet, no matter the mode. Use
   * this for items like diagnostics that you want ran during disabled,
   * autonomous, teleoperated and test.
   *
   * <p>This runs after the mode specific periodic functions, but before
   * LiveWindow and SmartDashboard integrated updating.
   */
  @Override
  public void robotPeriodic() {
  }

  /**
   * This autonomous (along with the chooser code above) shows how to select
   * between different autonomous modes using the dashboard. The sendable
   * chooser code works with the Java SmartDashboard. If you prefer the
   * LabVIEW Dashboard, remove all of the chooser code and uncomment the
   * getString line to get the auto name from the text box below the Gyro
   *
   * <p>You can add additional auto modes by adding additional comparisons to
   * the switch structure below with additional strings. If using the
   * SendableChooser make sure to add them to the chooser code above as well.
   */
  @Override
  public void autonomousInit() {
    elevator.ElevatorInit();
    isElevGoing = true;
    m_autoSelected = m_chooser.getSelected();
    // m_autoSelected = SmartDashboard.getString("Auto Selector", kDefaultAuto);
    System.out.println("Auto selected: " + m_autoSelected);
  }

  /**
   * This function is called periodically during autonomous.
   */
  @Override
  public void autonomousPeriodic() {
    teleopPeriodic();
  }
/**
   * This function is called once each time the robot enters teleoperated mode.
   */
  @Override
  public void teleopInit() 
  {
    if(!isElevGoing){
      elevator.ElevatorInit();
      isElevGoing = true;
    }
  }

  /**
   * This function is called periodically during operator control.
   */
  @Override
  public void teleopPeriodic() {
    long elevatorPosition = elevator.GetElevatorPosition();
    driveTrain.CalculateSwerve( DriverController, elevatorPosition );
    elevator.TestLiftPosition(ElevatorController);
    elevator.TestArmLift(ElevatorController);
    elevator.TestIntake(ElevatorController);
    elevator.ManageRumble(ElevatorController);
  }

  /**
   * This function is called periodically during test mode.
   */
  @Override
  public void testPeriodic() {
  }
}

File: 2026/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"

#include <frc2/command/CommandScheduler.h>

Robot::Robot() 
{
}

void Robot::RobotPeriodic() {
    
    m_timeAndJoystickReplay.Update();
    frc2::CommandScheduler::GetInstance().Run();
    

    m_chooser.SetDefaultOption(AvaliablePathPlannerAutos[0],AvaliablePathPlannerAutos[0]);
    for(auto Path : AvaliablePathPlannerAutos)
    {
        m_chooser.AddOption(Path, Path);
    }
    frc::SmartDashboard::PutData("Auto Modes", &m_chooser);

}

void Robot::DisabledInit() {}

void Robot::DisabledPeriodic() 
{}

void Robot::DisabledExit() {}

void Robot::AutonomousInit() {

    this->SetRobotAutoRoutine(m_chooser.GetSelected());
    m_autonomousCommand = this->GetRobotAutoCommand();


    if (m_autonomousCommand) {
        frc2::CommandScheduler::GetInstance().Schedule(std::move(m_autonomousCommand.value()));
    }
}

void Robot::AutonomousPeriodic() {}

void Robot::AutonomousExit() 
{
    frc2::CommandScheduler::GetInstance().CancelAll();
}

void Robot::TeleopInit() {
    // if (m_autonomousCommand) {
    //     frc2::CommandScheduler::GetInstance().Cancel(std::move(m_autonomousCommand.value()));
    // }
}

void Robot::TeleopPeriodic() {}

void Robot::TeleopExit() {}

void Robot::TestInit() {
    // frc2::CommandScheduler::GetInstance().CancelAll();
}

void Robot::TestPeriodic() {}

void Robot::TestExit() {}

#ifndef RUNNING_FRC_TESTS
int main() {
    return frc::StartRobot<Robot>();
}
#endif
